Overhaul cudf-polars docs for new streaming multi-GPU engines (#22252)
rapidsai/cudf: docs/cudf/source/cudf_polars/options.md

* `executor`: `CUDF_POLARS__EXECUTOR__<OPTION_NAME>` (e.g. `CUDF_POLARS__EXECUTOR__FALLBACK_MODE`)
* `engine`: `CUDF_POLARS__<OPTION_NAME>` (e.g. `CUDF_POLARS__RAISE_ON_FAIL`; nested prefixes for structured options)

### Category: `rapidsmpf`

See the [streaming runtime configuration reference][rapidsmpf-config] for the full list of fields and defaults.

### Category: `executor`

| Field                    | Description                                                                                                         | Default                      |
|--------------------------|---------------------------------------------------------------------------------------------------------------------|------------------------------|
| `num_py_executors`       | Workers for the internal Python `ThreadPoolExecutor`.                                                                                                              | `8`                          |
| `fallback_mode`          | When an unsupported operation forces a fallback to CPU execution: `"warn"`, `"raise"`, `"silent"`.                                                                 | `"warn"`                     |

> | `max_rows_per_partition` | Maximum number of rows per partition.                                                                                                                              | `1_000_000`                  |

Reviewer: ```suggestion
| `max_rows_per_partition` | Maximum number of rows per partition. Only used for in-memory `DataFrame` sources (never for disk IO or dynamic planning).                                                                                                                              | `1_000_000`                  |
```

> | `broadcast_join_limit`   | Maximum number of partitions for broadcast joins.                                                                                                                  | auto                         |

Reviewer: ```suggestion
| `broadcast_limit`   | Maximum number of bytes for broadcast joins.                                                                                                                  | auto                         |
```

> | `target_partition_size`  | Target I/O partition size in bytes. `0` means auto.                                                                                                                | auto                         |

Reviewer: ```suggestion
| `target_partition_size`  | Target partition size in bytes. Used for IO and dynamic planning. `0` means auto.                                                                                                                | auto                         |
```

|--------------------------|-------------------------------------------------------------------------------------------------------------------------------|---------------------------|
| `raise_on_fail`          | Raise an error instead of falling back to CPU execution.                                                                      | `False`                   |
| `parquet_options`        | Parquet configuration, dict or {class}`~cudf_polars.utils.config.ParquetOptions`.                                             | —                         |
| `memory_resource_config` | RMM configuration, dict or {class}`~cudf_polars.utils.config.MemoryResourceConfig`.                                           | —                         |
| `cuda_stream_policy`     | CUDA stream policy (`"default"`, `"pool"`, or a configuration dict).                                                          | —                         |
| `hardware_binding`       | Hardware binding policy. Pass a {class}`~cudf_polars.engine.hardware_binding.HardwareBindingPolicy` for fine-grained control. | `HardwareBindingPolicy()` |

> | `allow_gpu_sharing`      | When `False` (default), the engine raises if multiple ranks share the same physical GPU.                                      | `False`                   |

Reviewer 2: Can we reorder some of this. I think we _hope_ rapidsmpf options are less important?

opts = StreamingOptions(
    num_streaming_threads=8,
    fallback_mode="silent",
    spill_device_limit="70%",
)

with RayEngine.from_options(opts) as engine:
    result = (
        pl.scan_parquet("/data/*.parquet")
          .filter(pl.col("amount") > 100)
          .group_by("customer_id")
          .agg(pl.col("amount").sum())
          .collect(engine=engine)
    )
```

{class}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions` covers three
categories of fields:

| Category    | Scope                                                                                  | Env var prefix             |
| ----------- | -------------------------------------------------------------------------------------- | -------------------------- |
| `rapidsmpf` | RapidsMPF runtime, e.g. threads, CUDA streams, spilling, pinned memory, log level      | `RAPIDSMPF_`              |
| `executor`  | Query execution and partitioning, e.g. `max_rows_per_partition`, `fallback_mode`, ...  | `CUDF_POLARS__EXECUTOR__` |
| `engine`    | `pl.GPUEngine` kwargs, e.g. Parquet, memory resource, CUDA streams, hardware binding   | `CUDF_POLARS__`           |

> The `engine` category surfaces the same knobs as plain `pl.GPUEngine(...)` — for example,

Reviewer 2: ```suggestion
The `engine` category surfaces the same tuning knobs as plain `pl.GPUEngine(...)`. For example,
```

`parquet_options` and `memory_resource_config`. Configure these settings through
{class}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions` rather than
passing them to `pl.GPUEngine(...)` directly.

> The `rapidsmpf` category adds knobs for the streaming runtime that have no equivalent on the plain

Reviewer 2: ```suggestion
The `rapidsmpf` category adds configuration for the streaming runtime that have no equivalent on the plain
```

## In-memory Engine Options

The in-memory GPU engine is configured by passing keyword arguments to
[`polars.GPUEngine`][polars-gpuengine]:

```python
import polars as pl

engine = pl.GPUEngine(parquet_options={"chunked": True})
```

See the [Polars GPU support guide][polars-gpu] for the full in-memory usage story.

## Options Reference

Environment variables follow these patterns:

* `rapidsmpf`: `RAPIDSMPF_<OPTION_NAME>` (e.g. `RAPIDSMPF_NUM_STREAMING_THREADS`)
* `executor`: `CUDF_POLARS__EXECUTOR__<OPTION_NAME>` (e.g. `CUDF_POLARS__EXECUTOR__FALLBACK_MODE`)
* `engine`: `CUDF_POLARS__<OPTION_NAME>` (e.g. `CUDF_POLARS__RAISE_ON_FAIL`; nested prefixes for structured options)

### Category: `rapidsmpf`

See the [RapidsMPF configuration reference][rapidsmpf-config] for the full list of fields and defaults.

### Category: `executor`

| Field                    | Description                                                                                                         | Default                      |
| ------------------------ | ------------------------------------------------------------------------------------------------------------------- | ---------------------------- |

> | `num_py_executors`       | Number of workers for the internal Python `ThreadPoolExecutor`.                                                     | `8`                          |

Reviewer 2: I note that this is actually not the number of python threads that are used! We just get some random value depending on how asyncio decides to provide a thread pool (I keep on meaning to fix this...)

> | `fallback_mode`          | Fallback behavior (`"warn"`, `"raise"`, `"silent"`).                                                                | `"warn"`                     |

Reviewer 2: I don't know what "fallback behavior" means.

| `max_rows_per_partition` | Maximum number of rows per partition.                                                                               | `1_000_000`                  |
| `broadcast_join_limit`   | Maximum number of partitions for broadcast joins.                                                                   | auto                         |
| `target_partition_size`  | Target I/O partition size in bytes. `0` means auto.                                                                 | auto                         |
| `dynamic_planning`       | Dynamic planning configuration, dict or {class}`~cudf_polars.utils.config.DynamicPlanningOptions`. `None` disables. | enabled                      |

> | `unique_fraction`        | Per-column uniqueness estimate (0–1).                                                                               | `{}` (i.e. `1.0` per column) |

Reviewer 2: @rjzamora can this go away?

{class}`~cudf_polars.experimental.rapidsmpf.frontend.spmd.SPMDEngine`).
Build one and pass it to `Engine.from_options()` to construct a
{class}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine`:

```python
import polars as pl
from cudf_polars.experimental.rapidsmpf.frontend.options import StreamingOptions
from cudf_polars.experimental.rapidsmpf.frontend.ray import RayEngine

opts = StreamingOptions(
    num_streaming_threads=8,
    fallback_mode="silent",
    spill_device_limit="70%",
)

with RayEngine.from_options(opts) as engine:
    result = (
        pl.scan_parquet("/data/*.parquet")
          .filter(pl.col("amount") > 100)
          .group_by("customer_id")
          .agg(pl.col("amount").sum())
          .collect(engine=engine)
    )
```

{class}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions` covers three
categories of fields:

| Category    | Scope                                                                                  | Env var prefix             |
| ----------- | -------------------------------------------------------------------------------------- | -------------------------- |

> | `rapidsmpf` | RapidsMPF runtime, e.g. threads, CUDA streams, spilling, pinned memory, log level      | `RAPIDSMPF_*`              |

Reviewer 3: We're describing what the prefix is, so remove the `*`.

| `executor`  | Query execution and partitioning, e.g. `max_rows_per_partition`, `fallback_mode`, ...  | `CUDF_POLARS__EXECUTOR__*` |
| `engine`    | `pl.GPUEngine` kwargs, e.g. Parquet, memory resource, CUDA streams, hardware binding   | `CUDF_POLARS__*`           |

The `engine` category surfaces the same knobs as plain `pl.GPUEngine(...)` — for example,
`parquet_options` and `memory_resource_config`. You do not need to set them separately when using

> {class}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions`.

Reviewer 3: "separately" is a bit unclear here, since it's not clear thing is separate from what other thing.

I think you're saying "set them in `StreamingOptions()` rather than `pl.GPUEngine()`, so let's just say that: "configure these settings through `StreamingOptions`.

> The `rapidsmpf` category adds streaming-runtime knobs that have no equivalent on the plain

Reviewer 3: ```suggestion
The `rapidsmpf` category adds knobs for the streaming runtime that have no equivalent on the plain
```

`pl.GPUEngine`. See the [RapidsMPF configuration reference][rapidsmpf-config] for the underlying
meaning of each `RAPIDSMPF_*` field.

Every option has a corresponding environment variable. When an option is not set explicitly, its
value is read from the environment variable if present; otherwise the underlying library applies

> its built-in default. Boolean environment variables accept `{"1", "true", "yes", "y"}` as true

Reviewer 3: Are these case sensitive?

Author: no

and `{"0", "false", "no", "n"}` as false.

> All fields default to an `UNSPECIFIED` sentinel with the same precedence as the low-level

Reviewer 3: This paragraph feels a bit to implementation detail-y to me.

options: explicit value → environment variable → built-in default. Calling
`Engine.from_options(opts)` splits the fields into three dictionaries
(`to_rapidsmpf_options()`, `to_executor_options()`, `to_engine_options()`) and routes each to the
right destination.

## Building from a dictionary

{meth}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions.from_dict` accepts a
flat dict of field names. Unknown keys raise `TypeError`; `None` values leave the field
unspecified:

```python
opts = StreamingOptions.from_dict({
    "num_streaming_threads": 8,
    "fallback_mode": "silent",
})
```

This is convenient when options come from a config file or CLI.

> ## When to use StreamingOptions vs. raw keyword arguments

Reviewer 3: Bit of an odd section header. Maybe "Engine keyword arguments" instead?

And IIUC, `Engine` is a stand-in for `RayEngine`, `DaskEngine`, or `SPMDEngine`. Let's state that like "Each engine (`RayEngine`, `DaskEngine`, or `SPMDEngine`) accepts `rapidsmpf_options`, [...]. We recommend using this only when you need fine-grained control that doesn't fit the `StreamingOptions` schema. Otherwise, prefer using the engine's `from_options` constructor with `StreamingOptions`."

* Use {class}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions` +

>   `Engine.from_options(opts)` for typical streaming workflows — it is typed, centralized, and

Reviewer 3: ```suggestion
* Use {class}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions` with
  `Engine.from_options(opts)` for typical streaming workflows — it is typed, centralized, and
```

and maybe flip `Engine.from_options` and `StreamingOptions()`

  handles routing automatically.
* Use the raw `Engine(rapidsmpf_options=..., executor_options=..., engine_options=...)`
  constructor only when you need fine-grained control that doesn't fit the
  {class}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions` schema.
* For the in-memory engine,
  {class}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions` does not apply —
  pass keyword arguments to `pl.GPUEngine(...)` directly (see below).

> ## GPUEngine options

Reviewer 3: ```suggestion
## In-memory Engine Options
```

The in-memory GPU engine is configured by passing keyword arguments to
[`polars.GPUEngine`][polars-gpuengine]:

```python
import polars as pl

engine = pl.GPUEngine(parquet_options={"chunked": True})
```

See the [Polars GPU support guide][polars-gpu] for the full in-memory usage story.

> ## The options

Reviewer 3: ```suggestion
## Options Reference
```

Maybe?

Environment variables follow these patterns:

> * `rapidsmpf`: `RAPIDSMPF_<OPTION_NAME>`

Reviewer 3: Add one example for each. "e.g. `RAPIDSMPF_NUM_STREAMING_THREADS`".

* `executor`: `CUDF_POLARS__EXECUTOR__<OPTION_NAME>`
* `engine`: `CUDF_POLARS__<OPTION_NAME>` (or nested prefixes for structured options)

> ### Category: `rapidsmpf`

Reviewer 3: This feels liable to go out of date. Can we link to https://docs.rapids.ai/api/rapidsmpf/nightly/configuration/?
